kubeadm: Transition to using k8s.io/client-go

taimir opened the pull request:
**What this PR does / why we need it**: in `kubeadm`, replace all internal API clients with versioned clients from the `k8s.io/client-go` package.

**Which issue this PR fixes**: fixes https://github.com/kubernetes/kubernetes/issues/33914

**Special notes for your reviewer**:
I tried to use `k8s.io/client-go` everywhere where kubeadm uses some sort of API client to contact the API server.

For the certificate generation during a `join` command, resulting from a CSR, I had to copy over the  `RequestNodeCertificate` function from the `kubelet` code, as kubelet uses the internal API's and here I wanted to use the versioned `client-go` API client (otherwise the function signatures do not match). WDYT, is there a better way to do it right now?

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36425)
<!-- Reviewable:end -->

Commit:
kubeadm: Transition to using k8s.io/client-go

Replace all internal api clients with versioned clients from
k8s.io/client-go.

taimir commented on cmd/kubeadm/app/master/apiclient.go:
```diff
@@ -103,6 +104,15 @@ func CreateClientAndWaitForAPI(adminConfig *clientcmdapi.Config) (*clientset.Cli
 	return client, nil
 }
 
+func isNodeReady(node *api.Node) bool {
```
I find it strange that there is no such utility function in the `api/v1` api client, is there a reason for this? I can't use the one from the internal `api` as the `Node` in the signature differs.

luxas replied:
Just seems like `pkg/api/v1` doesn't have a `resource_helpers.go` file that `pkg/api` has.
We should consider adding `pkg/api/v1/resource_helpers.go`, yes

luxas replied:
Please add a TODO here to fix that

taimir commented on cmd/kubeadm/app/node/csr.go:
```diff
@@ -55,3 +62,71 @@ func PerformTLSBootstrap(connection *ConnectionDetails) (*clientcmdapi.Config, e
 	return finalConfig, nil
 }
+
+func RequestNodeCertificate(client certclient.CertificateSigningRequestInterface, privateKeyData []byte, nodeName types.NodeName) (certData []byte, err error) {
```
This is the function from the kubelet `csr` sub-package that I had to copy over to make it work with the `api/v1` client from `k8s.io/client-go`. Can you think of a better option right now?

luxas replied:
Instead add a TODO to `csr.RequestNodeCertificate` that it should use the versioned client and also a TODO to `RequestNodeCertificate()` in `pkg/kubelet/util/csr` to fix this.

dims commented:
@k8s-bot ok to test

luxas commented:
@taimir 
```
W1109 12:54:56.420] cmd/kubeadm/app/master/apiclient.go:183: undefined: "k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/api/unversioned".NodeLabelKubeadmAlphaRole
W1109 12:54:56.420] cmd/kubeadm/app/master/apiclient.go:183: undefined: "k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/api/unversioned".NodeLabelRoleMaster
W1109 12:54:56.421] cmd/kubeadm/app/master/apiclient.go:237: undefined: "k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/api/unversioned".NodeLabelKubeadmAlphaRole
W1109 12:54:56.421] cmd/kubeadm/app/master/apiclient.go:239: undefined: "k8s.io/kubernetes/vendor/k8s.io/client-go/pkg/api/unversioned".NodeLabelRoleMaster
```

Seems like you are using the unversioned client in client-go here, why?

luxas reviewed (commented):
Overall this looks good to me, thank you for doing this.

I'd like to hear from @kubernetes/sig-api-machinery what they think in these questions. My understanding from discussions on KubeCon was that all code in the repo should be switched over to versioned client-go.

taimir commented:
@luxas The constants `NodeLabelKubeadmAlphaRole`, `NodeLabelRoleMaster`, etc. are, rather surprisingly, contained only in the unversioned client packages (both in `k8s/pkg/api` and in `client-go`, in a file `well_known_labels.go`). The `kubeadm` code was using `k8s/pkg/api/unversioned` for those already before my PR, so I just converted it to `client-go/pkg/api/unversioned`.

I am not familiar with the guidelines used for structuring the different api clients, but there seems to be a pattern of constants & utility functions scattered at different places, when they could be used by all clients (internal and versioned ones). 

This results in the problem with the above mentioned role labels, as well as [here](https://github.com/kubernetes/kubernetes/pull/36425/files#diff-8562ce2534384a1bba24e11e71f60cb2R261), for example, where I could ditch the dependency on the `internal_api` altogether if I had another way to access the constant `NamespaceSystem` ('kube-system'). Same goes for the `resource_helpers.go` file from the discussion above. 

Are those points a subject of change? Should I just leave `TODOs` in my code for later?

luxas commented:
@taimir Please rebase this PR in order to fix the failures and leave TODOs wherever the code depends on constants or code outside of kubeadm that uses an unversioned client (like `RequestNodeCertificate`)
Please add TODOs on all places that are affecting kubeadm where an unversioned client is used.

I'd like to get this compiling and working, and then continue with the rest of the api refactoring.
Thanks

liggitt commented:
I know @caesarxuchao is working on a switch to client-go to drop right after the freeze lifts. Not sure if this area was in scope for that or not.

taimir committed:
Remove copied RequestNodeCertificate() function.

Add TODOs in the kubeadm code where still an unversioned client is used.

taimir commented on cmd/kubeadm/app/node/temp_util.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+// TODO: remove this in favor of kubeadmutil.WriteKubeconfigIfNotExists() when csr.RequestNodeCertificate() is fixed
+func WriteKubeconfigIfNotExists(name string, kubeconfig *clientcmdapi.Config) error {
```
@luxas I removed the `RequestNodeCertificate()` function that I had copied from `kubelet/util/csr` and went back to using an unversioned client for the CSR. However, since the CSR (and the client creation for the CSR) uses a bunch of utility functions from `kubeadm/app/util/kubeconfig.go`, I was forced to create copies again, because now those utilities functions are shared between places where a version client is used, and the CSR that uses a non-versioned one.

This is terribly ugly again, I cannot just change the signatures of the `kubeadmutil` functions just because the CSR uses an unversioned client, as those are used in other places too - that would mean going back to not using `client-go` anywhere.

WDYT? On purpose I have this change in a second commit in case we want to revert it ...

taimir committed:
Add TODO comments for constants from non-versioned

Constants and utility functions from non-versioned clients
should also be available in all versioned ones.

taimir commented:
What's up with the failures? I've already rebased ...

caesarxuchao commented:
https://github.com/kubernetes/kubernetes/pull/36673 converted kubeadm (and many other components) to use versioned clients.

caesarxuchao commented:
> For the certificate generation during a join command, resulting from a CSR, I had to copy over the RequestNodeCertificate function from the kubelet code, as kubelet uses the internal API's and here I wanted to use the versioned client-go API client (otherwise the function signatures do not match). 

#36673 also converted kubelet to use versioned API, so the problem didn't exist there.

luxas commented:
@caesarxuchao So do you mean we should close this in favor for #36673?

caesarxuchao commented:
@luxas yes, sorry.

luxas commented:
Sorry @taimir I really appreciate your work.
Hopefully you can find something other to work on since we don't want to cause unneccesary rebases/conflicts with @caesarxuchao

taimir commented:
No problem guys, glad to see this is fixed :)
